oeps/oep-0013.rst

```diff
@@ -0,0 +1,576 @@
+Use HTTP verbs to operate on the collections and resources.
+
+**"Safe"** operations (*Must not*: modify any resource's state)
+
+- ``GET`` - Filter a collection down to a set of resources matching the provided criteria ::
+
+    /users
+    /users/:id
+    /users?last_name=Smith
+
+**State-modyfing** operations
+
+- ``POST`` - Submit information about a resource to the service for processing
+
+    - Typically utilized for resource creation workflows ::
+
+        POST /users   {user_data}
+
+    - Other cases include establishing a link between two resources ::
+
+        POST /users/:id/groups  {group_id}
+
+- ``PATCH`` - Modify an existing resource
+
+    - Return value must be 204 with no additional content. 
+    - If implementing the simpler `"merge patch" algorithm <https://tools.ietf.org/html/rfc7396>`_, the content_type of ``application/merge-patch+json`` should be used to differentiate from the full `JSON Patch specification <http://tools.ietf.org/html/rfc6902>`_.
+    - See the note below for more information on ``PATCH``.
+
+- ``PUT`` - Replace a resource in entirety.  
+
+    - Most use cases can be solved using partial updates via ``PATCH``
+    - ``PUT`` can be used if a complete update of the resource, including all of its sub-resources, is desired. 
+    - A typical ``PUT`` use case:  The resource is a singular primitive value addressable via URI, such as when updating the value for a particular User Preference.
+
+- ``DELETE`` - Remove a resource (or a relationship) from the system
+
+    - **Must** return ``HTTP 204 No Content`` ::
+
+        DELETE /users/:id
+        DELETE /users/:id/groups/:group_id
+
+**A note on HTTP PATCH**
+
+"Plain" HTTP PATCH (RFC 5789) is neutral with respect to content type and only specifies that the request body provide instructions for how to update the resource in question, not the format of those instructions.  Several flavors of PATCH specific to JSON documents have more explicit definition, as noted above.
+
+
+In DRF, the default PATCH handling fairly closely resembles what is specified by merge patch, but it does not require (or understand) the ``application/merge-patch+json`` media type.  In edX REST APIs, if an implementation will use DRF's default PATCH handling, the implementation MUST recognize and accept the merge patch media type; API clients SHOULD use this media type, preferring it to the more typical ``application/json`` type.
+
+The reason for this stricture is to explicitly leave room for supporting multiple PATCH styles in any given API, should this become desirable, without breaking existing clients.
+
+URL parameters
+--------------
+
+- *Must*: Sweep complexity behind the ``?`` ::
+
+    GET /dogs?color=red&state=running&location=park
+
+- ``fields`` parameter - Allow clients to specify/filter the fields in the response by supporting a fields parameter as a comma-delimited list. ::
+
+    /dogs?fields=name,color,location
+    /dogs?fields=title,media:group(media:thumbnail)
```

Comment:
1. I don't think we have implemented the xpathesque type of field filtering that Google has.
2. If we think we want to use this, for mobile or otherwise, we should choose a solution and duplicate and/or link to docs that provide details.  For example, [Google's docs (drive api) for partial responses](https://developers.google.com/drive/v3/web/performance#partial-response).